packages/ipfs-message-port-client/src/core.js

```diff
@@ -0,0 +1,479 @@
+      : undefined
+
+    const result = await this.remote.addAll({
+      ...options,
+      input: encodeAddAllInput(input, transfer),
+      progress,
+      transfer,
+      timeout,
+      signal
+    })
+    yield * decodeIterable(result.data, decodeAddedData)
+  }
+
+  /**
+   * Add file to IPFS.
+   *
+   * If you pass binary data like `Uint8Array` it is recommended to provide
+   * `transfer: [input.buffer]` which would allow transferring it instead of
+   * copying.
+   *
+   * @param {AddInput} input
+   * @param {Object} [options]
+   * @param {string} [options.chunker="size-262144"]
+   * @param {number} [options.cidVersion=0]
+   * @param {boolean} [options.enableShardingExperiment]
+   * @param {string} [options.hashAlg="sha2-256"]
+   * @param {boolean} [options.onlyHash=false]
+   * @param {boolean} [options.pin=true]
+   * @param {function(number):void} [options.progress]
+   * @param {boolean} [options.rawLeaves=false]
+   * @param {number} [options.shardSplitThreshold=1000]
+   * @param {boolean} [options.trickle=false]
+   * @param {boolean} [options.wrapWithDirectory=false]
+   * @param {number} [options.timeout]
+   * @param {Transferable[]} [options.transfer]
+   * @param {AbortSignal} [options.signal]
+   * @returns {Promise<AddedData>}
+   */
+  async add (input, options = {}) {
+    const { timeout, signal } = options
+    const transfer = [...(options.transfer || [])]
+    const progress = options.progress
+      ? encodeCallback(options.progress, transfer)
+      : undefined
+
+    const result = await this.remote.add({
+      ...options,
+      input: encodeAddInput(input, transfer),
+      progress,
+      transfer,
+      timeout,
+      signal
+    })
+
+    return decodeAddedData(result.data)
+  }
+
+  /**
+   * Returns content addressed by a valid IPFS Path.
+   * @param {string|CID} inputPath
+   * @param {Object} [options]
+   * @param {number} [options.offset]
+   * @param {number} [options.length]
+   * @param {number} [options.timeout]
+   * @param {AbortSignal} [options.signal]
+   * @returns {AsyncIterable<Uint8Array>}
+   */
+  async * cat (inputPath, options = {}) {
+    const input = CID.isCID(inputPath) ? encodeCID(inputPath) : inputPath
+    const result = await this.remote.cat({ ...options, path: input })
+    yield * decodeIterable(result.data, identity)
+  }
+}
+
+/**
+ * Decodes values yield by `ipfs.add`.
+ * @param {AddedEntry} data
+ * @returns {AddedData}
+ */
+const decodeAddedData = ({ path, cid, mode, mtime, size }) => {
+  return {
+    path,
+    cid: decodeCID(cid),
+    mode,
+    mtime,
+    size
+  }
+}
+
+/**
+ * @template T
+ * @param {T} v
+ * @returns {T}
+ */
+const identity = (v) => v
+
+/**
+ * Encodes input passed to the `ipfs.add` via the best possible strategy for the
+ * given input.
+ *
+ * @param {AddInput} input
+ * @param {Transferable[]} transfer
+ * @returns {EncodedAddInput}
+ */
+const encodeAddInput = (input, transfer) => {
```

Comment:
It would remove first 4 conditions, but the at the cost of: 

1. `normaliseInput` will turn everything into `Blob`, so things that could have being transferred  would be copied.
2. Introduce artificial asynchrony.
3. TS will no longer be helpful here because `normaliseInput` is untyped.

Considering cost / benefit ration I would much rather keep it as is & attempt to unify two as the followup.